src/server/admin/server/api_server.go

```diff
@@ -335,43 +363,204 @@ func (a *apiServer) Restore(restoreServer admin.API_RestoreServer) (retErr error
+			return fmt.Errorf("error creating repo: %v", grpcutil.ScrubGRPC(err))
+		}
+	case op.Commit != nil:
+		if _, err := pachClient.PfsAPIClient.BuildCommit(pachClient.Ctx(), op.Commit); err != nil && !errutil.IsAlreadyExistError(err) {
+			return fmt.Errorf("error creating commit: %v", grpcutil.ScrubGRPC(err))
+		}
+	case op.Branch != nil:
+		if op.Branch.Branch == nil {
+			op.Branch.Branch = client.NewBranch(op.Branch.Head.Repo.Name, op.Branch.SBranch)
+		}
+		if _, err := pachClient.PfsAPIClient.CreateBranch(pachClient.Ctx(), op.Branch); err != nil && !errutil.IsAlreadyExistError(err) {
+			return fmt.Errorf("error creating branch: %v", grpcutil.ScrubGRPC(err))
+		}
+	case op.Pipeline != nil:
+		if op.Pipeline.Salt != "" {
+			op.Pipeline.Salt = "" // clear salt so we don't re-use old datum hashtrees
+		}
+		if _, err := pachClient.PpsAPIClient.CreatePipeline(pachClient.Ctx(), op.Pipeline); err != nil && !errutil.IsAlreadyExistError(err) {
+			return fmt.Errorf("error creating pipeline: %v", grpcutil.ScrubGRPC(err))
+		}
+	}
+	return nil
+}
+
+func (a *apiServer) apply1_7Op(pachClient *client.APIClient, op *admin.Op1_7) error {
+	switch {
+	case op.Tag != nil:
+		newTagObjectRequest := &pfs.TagObjectRequest{
+			Object: convert1_7Object(op.Tag.Object),
+			Tags:   convert1_7Tags(op.Tag.Tags),
+		}
+		if _, err := pachClient.ObjectAPIClient.TagObject(
+			pachClient.Ctx(),
+			newTagObjectRequest,
+		); err != nil {
+			return fmt.Errorf("error tagging object: %v", grpcutil.ScrubGRPC(err))
+		}
+	case op.Repo != nil:
+		newCreateRepoRequest := &pfs.CreateRepoRequest{
+			Repo:        convert1_7Repo(op.Repo.Repo),
+			Description: op.Repo.Description,
+		}
+		if _, err := pachClient.PfsAPIClient.CreateRepo(
+			pachClient.Ctx(),
+			newCreateRepoRequest,
+		); err != nil && !errutil.IsAlreadyExistError(err) {
+			return fmt.Errorf("error creating repo: %v", grpcutil.ScrubGRPC(err))
+		}
+	case op.Commit != nil:
+		// update hashtree
+		var buf bytes.Buffer
+		if err := a.pachClient.GetObject(op.Commit.Tree.Hash, &buf); err != nil {
+			return err
+		}
+		var oldTree hashtree_1_7.HashTreeProto
+		oldTree.Unmarshal(buf.Bytes())
+		newTree, err := convert1_7HashTree(a.storageRoot, &oldTree)
+		if err != nil {
+			return err
+		}
+		defer newTree.Destroy()
+
+		// write new hashtree as an object
+		w, err := pachClient.PutObjectAsync(nil)
+		if err != nil {
+			return fmt.Errorf("could not put new hashtree for commit %q: %v", op.Commit.ID, err)
+		}
+		newTree.Serialize(w)
+		if err := w.Close(); err != nil {
+			return fmt.Errorf("could finish object containing new hashtree for commit %q: %v", op.Commit.ID, err)
+		}
+		newTreeObj, err := w.Object()
+		if err != nil {
+			return fmt.Errorf("could retrieve object reference to new hashtree for commit %q: %v", op.Commit.ID, err)
+		}
+
+		// Set op's object to new hashtree & finish building commit
+		newBuildCommitRequest := &pfs.BuildCommitRequest{
+			Parent:     convert1_7Commit(op.Commit.Parent),
```

Comment:
Do we need convert methods for things that haven't changed? Looks like the codes already written, but that's going to be a pain to write each time. Can we do some sort of proto based conversion?

Reply:
Just to update, reflection, isn't working well for this as I'd hoped, because I keep having to convert types back and forth.

I think one thing worth considering here is that this code is converting 1.7 protos, and those can't change anymore, so it's not like we'll have much maintenance to do here (and if we add new fields to the output protos that have to be set, automatic name-based conversion wouldn't work for the new fields anyway, and we'd still have to write custom code)

Reply:
Yeah, per offline discussion let's not do this. If this is sufficiently annoying for new migrations I'm sure the engineer tasked with doing it will have every incentive they need to come up with something simpler.